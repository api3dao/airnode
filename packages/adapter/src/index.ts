export { buildRequest, executeRequest, buildAndExecuteRequest } from './request-building';

export {
  encodeValue,
  extractValue,
  extractAndEncodeResponse,
  multiplyValue,
  bigNumberToString,
} from './response-processing';

export * from './clients';
export * from './types';
