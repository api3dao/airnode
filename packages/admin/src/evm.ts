import { ethers } from 'ethers';
import { AirnodeRrpArtifact, AirnodeRrpAddresses } from '@airnode/protocol';

export async function getAirnodeRrp(providerUrl: string) {
  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
  const network = await provider.getNetwork();
  return new ethers.Contract(
    AirnodeRrpAddresses[network.chainId]! /** TODO: shouldn't this be defined? */,
    AirnodeRrpArtifact.abi,
    provider
  );
}

export async function getAirnodeRrpWithSigner(mnemonic: string, providerUrl: string) {
  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
  const wallet = ethers.Wallet.fromMnemonic(mnemonic).connect(provider);
  const network = await provider.getNetwork();
  return new ethers.Contract(
    AirnodeRrpAddresses[network.chainId]! /** TODO: shouldn't this be defined? */,
    AirnodeRrpArtifact.abi,
    wallet
  );
}
