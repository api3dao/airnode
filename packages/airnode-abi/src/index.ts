export { encode } from './encoding';
export { decode } from './decoding';
export { PARAMETER_SHORT_TYPES } from './constants';
export * from './types';
