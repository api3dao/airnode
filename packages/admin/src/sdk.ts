import { AirnodeRrp } from '@airnode/protocol';
import * as evm from './evm';
import * as admin from '.';

/*
 * Class versin of the SDK bound to AirnodeRrp contract instance
 */
export class AdminSdk {
  static getAirnodeRrp = evm.getAirnodeRrp;
  static getAirnodeRrpWithSigner = evm.getAirnodeRrpWithSigner;

  constructor(public airnodeRrp: AirnodeRrp) {}

  clientAddressToNoRequests = (clientAddress: string) =>
    admin.clientAddressToNoRequests(this.airnodeRrp, clientAddress);

  // TODO: create other methods
}
