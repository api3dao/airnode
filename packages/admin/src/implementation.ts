import { ethers } from 'ethers';
import * as airnodeAbi from '@api3/airnode-abi';
import { AirnodeRrp } from '@api3/protocol';

type Last<T extends any[]> = T extends [...infer _, infer L] ? L : never;
const last = <T extends any[]>(array: T): Last<T> => array[array.length - 1];

const assertAllParamsAreReturned = (params: object, ethersParams: any[]) => {
  if (Object.keys(params).length !== ethersParams.length) {
    throw new Error(`SDK doesn't return some of the parameters!`);
  }
};

/**
 * HD wallets allow us to create multiple accounts from a single mnemonic.
 * Each requester creates a designated wallet for each provider to use
 * in order for them to be able to respond to the requests their clients make.
 *
 * By convention derivation paths start with a master index
 * followed by child indexes that can be any integer up to 2^31.
 *
 * Since addresses can be represented as 160bits (20bytes) we can then
 * split it in chunks of 31bits and create a path with the following pattern:
 * m/0/1st31bits/2nd31bits/3rd31bits/4th31bits/5th31bits/6th31bits.
 *
 * @param address A string representing a 20bytes hex address
 * @returns The path derived from the address
 */
export const addressToDerivationPath = (address: string): string => {
  const requesterBN = ethers.BigNumber.from(ethers.utils.getAddress(address));
  const paths = [];
  for (let i = 0; i < 6; i++) {
    const shiftedRequesterBN = requesterBN.shr(31 * i);
    paths.push(shiftedRequesterBN.mask(31).toString());
  }
  return paths.join('/');
};

export async function deriveDesignatedWallet(airnodeRrp: AirnodeRrp, airnodeId: string, requester: string) {
  const airnode = await airnodeRrp.getAirnodeParameters(airnodeId);
  const hdNode = ethers.utils.HDNode.fromExtendedKey(airnode.xpub);
  const derivationPath = addressToDerivationPath(requester);
  const designatedWalletNode = hdNode.derivePath(`m/0/${derivationPath}`);
  return designatedWalletNode.address;
}

export async function endorseClient(airnodeRrp: AirnodeRrp, clientAddress: string) {
  await airnodeRrp.setClientEndorsementStatus(clientAddress, true);
  return clientAddress;
}

export async function unendorseClient(airnodeRrp: AirnodeRrp, clientAddress: string) {
  await airnodeRrp.setClientEndorsementStatus(clientAddress, false);
  return clientAddress;
}

export interface Template {
  parameters: string | airnodeAbi.InputParameter[];
  airnodeId: string;
  endpointId: string;
}

export async function createTemplate(airnodeRrp: AirnodeRrp, template: Template) {
  let encodedParameters;
  if (typeof template.parameters == 'string') {
    encodedParameters = template.parameters;
  } else {
    encodedParameters = airnodeAbi.encode(template.parameters);
  }
  const tx = await airnodeRrp.createTemplate(template.airnodeId, template.endpointId, encodedParameters);

  return new Promise<string>((resolve) =>
    airnodeRrp.provider.once(tx.hash, ({ logs }) => {
      const parsedLog = airnodeRrp.interface.parseLog(logs[0]);
      resolve(parsedLog.args.templateId);
    })
  );
}

export async function requestWithdrawal(
  airnodeRrp: AirnodeRrp,
  airnodeId: string,
  requester: string,
  destination: string
) {
  const designatedWalletAddress = await deriveDesignatedWallet(airnodeRrp, airnodeId, requester);
<<<<<<< HEAD
  const tx = await airnodeRrp.requestWithdrawal(airnodeId, designatedWalletAddress, destination);
=======
  await airnodeRrp.requestWithdrawal(airnodeId, designatedWalletAddress, destination);
  const filter = airnodeRrp.filters.WithdrawalRequested(null, null, null, null, null);
>>>>>>> ae70189b... Merge branch 'beta-protocol' into authorizer-tests-combined

  return new Promise<string>((resolve) =>
    airnodeRrp.provider.once(tx.hash, ({ logs }) => {
      const parsedLog = airnodeRrp.interface.parseLog(logs[0]);
      resolve(parsedLog.args.withdrawalRequestId);
    })
  );
}

export async function checkWithdrawalRequest(airnodeRrp: AirnodeRrp, requestId: string) {
  const filter = airnodeRrp.filters.WithdrawalFulfilled(null, null, requestId, null, null, null);

  const logs = await airnodeRrp.queryFilter(filter);
  if (logs.length === 0) {
    return undefined;
  }

  const ethersLogParams = logs[0].args;
  // remove array parameters from ethers response
  const { airnodeId, amount, designatedWallet, destination, requester, withdrawalRequestId } = ethersLogParams;
  const logParams = { airnodeId, amount, designatedWallet, destination, requester, withdrawalRequestId };
  assertAllParamsAreReturned(logParams, ethersLogParams);

  // cast ethers BigNumber for portability
  return { ...logParams, amount: amount.toString() };
}

const isEthersWallet = (signer: any): signer is ethers.Wallet => !!signer.mnemonic;

export async function setAirnodeParameters(airnodeRrp: AirnodeRrp, authorizers: string[]) {
  const wallet = airnodeRrp.signer;

  if (!isEthersWallet(wallet)) {
    throw new Error('Expected AirnodeRrp contract signer must be ethers.Wallet!');
  }

  const hdNode = ethers.utils.HDNode.fromMnemonic(wallet.mnemonic.phrase);
  const xpub = hdNode.neuter().extendedKey;
  const masterWallet = ethers.Wallet.fromMnemonic(wallet.mnemonic.phrase, 'm').connect(
    airnodeRrp.provider as ethers.providers.Provider
  );
  // Assuming masterWallet has funds to make the transaction below
<<<<<<< HEAD
  const tx = await airnodeRrp.connect(masterWallet).setAirnodeParameters(xpub, authorizers);
=======
  await airnodeRrp.connect(masterWallet).setAirnodeParameters(xpub, authorizers);
  const filter = airnodeRrp.filters.AirnodeParametersSet(null, null, null, null);
>>>>>>> ae70189b... Merge branch 'beta-protocol' into authorizer-tests-combined

  return new Promise<string>((resolve) =>
    airnodeRrp.provider.once(tx.hash, ({ logs }) => {
      const parsedLog = airnodeRrp.interface.parseLog(logs[0]);
      resolve(parsedLog.args.airnodeId);
    })
  );
}

export async function deriveEndpointId(oisTitle: string, endpointName: string) {
  return ethers.utils.keccak256(ethers.utils.defaultAbiCoder.encode(['string'], [`${oisTitle}_${endpointName}`]));
}

export async function clientAddressToNoRequests(airnodeRrp: AirnodeRrp, clientAddress: string) {
  return (await airnodeRrp.clientAddressToNoRequests(clientAddress)).toString();
}

export async function getAirnodeParameters(airnodeRrp: AirnodeRrp, airnodeId: string) {
  const ethersParams = await airnodeRrp.getAirnodeParametersAndBlockNumber(airnodeId);

  // remove array parameters from ethers response
  const { admin, xpub, authorizers, blockNumber } = ethersParams;
  const params = { admin, xpub, authorizers, blockNumber };
  assertAllParamsAreReturned(params, ethersParams);

  // cast ethers BigNumber for portability
  return { ...params, blockNumber: blockNumber.toString() };
}

export async function getTemplate(airnodeRrp: AirnodeRrp, templateId: string) {
  const ethersTemplate = await airnodeRrp.getTemplate(templateId);

  // remove array parameters from ethers response
  const { airnodeId, endpointId, parameters } = ethersTemplate;
  const template = { airnodeId, endpointId, parameters };
  assertAllParamsAreReturned(template, ethersTemplate);

  return template;
}

export async function getTemplates(airnodeRrp: AirnodeRrp, templateIds: string[]) {
  const ethersTemplate = await airnodeRrp.getTemplates(templateIds);

  // remove array parameters from ethers response
  const { airnodeIds, endpointIds, parameters } = ethersTemplate;
  const templates = { airnodeIds, endpointIds, parameters };
  assertAllParamsAreReturned(templates, ethersTemplate);

  const formattedTemplates = airnodeIds.map((_, i) => ({
    airnodeId: airnodeIds[i],
    endpointId: endpointIds[i],
    parameters: parameters[i],
  }));
  return formattedTemplates;
}

export function requesterToClientAddressToEndorsementStatus(
  airnodeRrp: AirnodeRrp,
  requester: string,
  clientAddress: string
) {
  return airnodeRrp.requesterToClientAddressToEndorsementStatus(requester, clientAddress);
}

export async function requesterToNextWithdrawalRequestIndex(airnodeRrp: AirnodeRrp, requester: string) {
  const requestsCount = await airnodeRrp.requesterToNextWithdrawalRequestIndex(requester);
  return requestsCount.toString();
}

export interface FulfillWithdrawalReturnValue {
  airnodeId: string;
  requester: string;
  designatedWallet: string;
  destination: string;
  amount: string;
  withdrawalRequestId: string;
}

export async function fulfillWithdrawal(
  airnodeRrp: AirnodeRrp,
  requestId: string,
  airnodeId: string,
  requester: string,
  destination: string,
  amount: string
) {
  await airnodeRrp.fulfillWithdrawal(requestId, airnodeId, requester, destination, {
    value: ethers.utils.parseEther(amount),
  });
  const filter = airnodeRrp.filters.WithdrawalFulfilled(airnodeId, requester, requestId, null, null, null);

  return new Promise<FulfillWithdrawalReturnValue>((resolve) =>
    airnodeRrp.once(filter, (...args) => {
      // remove array parameters from ethers response
      const { airnodeId, requester, designatedWallet, destination, amount, withdrawalRequestId } = last(args).args;
      const params = { airnodeId, requester, designatedWallet, destination, amount, withdrawalRequestId };

      // cast ethers BigNumber for portability
      resolve({ ...params, amount: amount.toString() });
    })
  );
}
