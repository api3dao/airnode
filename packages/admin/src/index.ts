import { ethers } from 'ethers';
import * as airnodeAbi from '@airnode/airnode-abi';

// TODO: add stronger types for airnodeRrp contract

export async function createRequester(airnodeRrp: ethers.Contract, requesterAdmin: string) {
  const receipt = await airnodeRrp.createRequester(requesterAdmin);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.requesterIndex.toString());
    })
  );
}

export async function setRequesterAdmin(airnodeRrp: ethers.Contract, requesterIndex: string, requesterAdmin: string) {
  const receipt = await airnodeRrp.setRequesterAdmin(requesterIndex, requesterAdmin);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.admin);
    })
  );
}

export async function deriveDesignatedWallet(airnodeRrp: ethers.Contract, airnodeId: string, requesterIndex: string) {
  const airnode = await airnodeRrp.getAirnodeParameters(airnodeId);
  const hdNode = ethers.utils.HDNode.fromExtendedKey(airnode.xpub);
  const designatedWalletNode = hdNode.derivePath(`m/0/${requesterIndex}`);
  return designatedWalletNode.address;
}

export async function endorseClient(airnodeRrp: ethers.Contract, requesterIndex: string, clientAddress: string) {
  const receipt = await airnodeRrp.setClientEndorsementStatus(requesterIndex, clientAddress, true);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.clientAddress);
    })
  );
}

export async function unendorseClient(airnodeRrp: ethers.Contract, requesterIndex: string, clientAddress: string) {
  const receipt = await airnodeRrp.setClientEndorsementStatus(requesterIndex, clientAddress, false);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.clientAddress);
    })
  );
}

// TODO: type me
interface Template {
  parameters: string | airnodeAbi.InputParameter[];
  airnodeId: string;
  endpointId: string;
}

export async function createTemplate(airnodeRrp: ethers.Contract, template: Template) {
  let encodedParameters;
  if (typeof template.parameters == 'string') {
    encodedParameters = template.parameters;
  } else {
    encodedParameters = airnodeAbi.encode(template.parameters);
  }
  const receipt = await airnodeRrp.createTemplate(template.airnodeId, template.endpointId, encodedParameters);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.templateId);
    })
  );
}

export async function requestWithdrawal(
  airnodeRrp: ethers.Contract,
  airnodeId: string,
  requesterIndex: string,
  destination: string
) {
  const designatedWalletAddress = deriveDesignatedWallet(airnodeRrp, airnodeId, requesterIndex);
  const receipt = await airnodeRrp.requestWithdrawal(airnodeId, requesterIndex, designatedWalletAddress, destination);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.withdrawalRequestId);
    })
  );
}

export async function checkWithdrawalRequest(airnodeRrp: ethers.Contract, withdrawalRequestId: string) {
  const logs = await airnodeRrp.provider.getLogs({
    address: airnodeRrp.address,
    fromBlock: 0,
    topics: [
      ethers.utils.id('WithdrawalFulfilled(bytes32,uint256,bytes32,address,address,uint256)'),
      // these are wrong typings by ethers.js: https://github.com/ethers-io/ethers.js/issues/1434
      null as any,
      null,
      withdrawalRequestId,
    ],
  });
  if (logs.length === 0) {
    return undefined;
  }
  const parsedLog = airnodeRrp.interface.parseLog(logs[0]!);
  return parsedLog.args.amount;
}

export async function setAirnodeParameters(
  airnodeRrp: ethers.Contract,
  airnodeAdmin: string,
  authorizers: object /** TODO: */
) {
  const wallet = airnodeRrp.signer as ethers.Wallet;
  const hdNode = ethers.utils.HDNode.fromMnemonic(wallet.mnemonic.phrase);
  const xpub = hdNode.neuter().extendedKey;
  const masterWallet = ethers.Wallet.fromMnemonic(wallet.mnemonic.phrase, 'm').connect(airnodeRrp.provider);
  // Assuming masterWallet has funds to make the transaction below
  const receipt = await airnodeRrp
    .connect(masterWallet)
    .setAirnodeParametersAndForwardFunds(airnodeAdmin, xpub, authorizers);
  return new Promise((resolve) =>
    airnodeRrp.provider.once(receipt.hash, (tx) => {
      const parsedLog = airnodeRrp.interface.parseLog(tx.logs[0]);
      resolve(parsedLog.args.airnodeId.toString());
    })
  );
}

export async function deriveEndpointId(oisTitle: string, endpointName: string) {
  return ethers.utils.keccak256(ethers.utils.defaultAbiCoder.encode(['string'], [`${oisTitle}_${endpointName}`]));
}
