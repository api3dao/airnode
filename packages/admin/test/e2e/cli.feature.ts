import { execSync } from 'child_process';
import { existsSync } from 'fs';
import difference from 'lodash/difference';
import { AirnodeRrp, AirnodeRrpFactory } from '@api3/protocol';
import { ethers } from 'ethers';
import * as admin from '../../src';

const PROVIDER_URL = 'http://127.0.0.1:8545/';
const CLI_EXECUTABLE = `${__dirname}/../../dist/cli.js`;
// Turning this flag to 'true' will print each command before executing it
// It might be useful to turn on, while debugging particular test.
const DEBUG_COMMANDS = false;

it('has disabled DEBUG_COMMANDS flag', () => {
  expect(DEBUG_COMMANDS).toBe(false);
});

describe('CLI', () => {
  let provider: ethers.providers.JsonRpcProvider;
  let deployer: ethers.providers.JsonRpcSigner;
  const aliceDerivationPath = "m/44'/60'/0'/0/1";
  let alice: ethers.Wallet;
  const bobDerivationPath = "m/44'/60'/0'/0/2";
  let bob: ethers.Wallet;
  let airnodeRrp: AirnodeRrp;
  // https://hardhat.org/hardhat-network/#hardhat-network-initial-state
  const mnemonic = 'test test test test test test test test test test test junk';

  type CommandArg = [string, string | string[]];
  const execCommand = (command: string, ...args: CommandArg[]) => {
    const quote = (val: string) => `"${val}"`;
    const formattedArgs = args
      .map(([c, a]) => {
        // if args is array then quote each elem and separate them with space
        if (Array.isArray(a)) return `${c} ${a.map(quote).join(' ')}`;
        // otherwise just quote each elem and separate them with space
        else return `${c} ${quote(a)}`;
      })
      .join(' ');
    const formattedCommand = `${command} ${formattedArgs}`;
    if (DEBUG_COMMANDS) console.log(`Executing command: ${formattedCommand}`);
    try {
      return execSync(`node ${CLI_EXECUTABLE} ${formattedCommand}`).toString().trim();
    } catch (e: any) {
      // rethrow the output of the CLI
      throw new Error(e.stdout.toString().trim());
    }
  };

  const deriveDesignatedWallet = async (wallet: ethers.Wallet, requester: string) => {
    const airnodeMnemonic = wallet.mnemonic.phrase;
    const derivationPath = admin.addressToDerivationPath(requester);
    return ethers.Wallet.fromMnemonic(airnodeMnemonic, `m/0/${derivationPath}`).connect(provider);
  };

  beforeAll(() => {
    jest.setTimeout(45_000);

    provider = new ethers.providers.JsonRpcProvider(PROVIDER_URL);
    deployer = provider.getSigner(0);
    alice = ethers.Wallet.fromMnemonic(mnemonic, aliceDerivationPath).connect(provider);
    bob = ethers.Wallet.fromMnemonic(mnemonic, bobDerivationPath).connect(provider);
  });

  beforeEach(async () => {
    airnodeRrp = await new AirnodeRrpFactory(deployer).deploy();
  });

  it('exposes the CLI executable', () => {
    expect(existsSync(`${CLI_EXECUTABLE}`)).toBe(true);
  });

  it('shows help', () => {
    const output = execCommand('--help');
    expect(output).toMatchSnapshot();
  });

  it('has command for every public function/mapping', () => {
    const camelToKebabCase = (str: string) => str.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`);

    const allFunctions = Object.keys(airnodeRrp.functions)
      .filter((key) => !key.includes('('))
      .map(camelToKebabCase);
    const allCommands = execCommand('--help')
      .split(/Commands|Options/)[1]
      .split(new RegExp('cli\\.js'))
      .slice(1)
      .map((line) => line.trim().split(' ')[0]);

    const sdkCliDiff = difference(allFunctions, allCommands);
    const uncoveredFunctions = [
      // renamed to get-airnode-parameters (which calls this more generic version)
      'get-airnode-parameters-and-block-number',
      // renamed to get-endorsement-status
      'requester-to-client-address-to-endorsement-status',
      // covered by endorse-client and unendrose-client commands
      'set-client-endorsement-status',
      // explicitely not implemented (not useful as of now)
      'get-templates',
      'check-authorization-status',
      'check-authorization-statuses',
      'default-authorizers',
      'fail',
      'fulfill',
      'fulfill-withdrawal',
      'make-full-request',
      'make-request',
      'owner',
      'renounce-ownership',
      'request-with-id-has-failed',
      'client-address-to-no-requests',
      'requester-to-next-withdrawal-request-index',
      'set-default-authorizers',
      'transfer-ownership',
    ].sort();
    expect(sdkCliDiff).toEqual(uncoveredFunctions);
  });

  it('derives the designated wallet', async () => {
    const requester = alice.address;
    const airnodeWallet = ethers.Wallet.createRandom().connect(provider);
    const airnodeMnemonic = airnodeWallet.mnemonic.phrase;
    const masterWallet = ethers.Wallet.fromMnemonic(airnodeMnemonic, 'm').connect(provider);
    // Fund the master wallet - which will be used to set the airnode parameters
    await deployer.sendTransaction({
      to: masterWallet.address,
      value: ethers.utils.parseEther('1'),
    });
    airnodeRrp = airnodeRrp.connect(airnodeWallet);
    const airnodeId = await admin.setAirnodeParameters(airnodeRrp, []);

    // Derive the wallet using CLI and admin SDK
    const out = execCommand(
      'derive-designated-wallet',
      ['--providerUrl', PROVIDER_URL],
      ['--airnodeRrp', airnodeRrp.address],
      ['--airnodeId', airnodeId],
      ['--requester', requester]
    );

    // Derive the wallet programatically
    const designatedWallet = await deriveDesignatedWallet(airnodeWallet, requester);

    // Check that they generate the same wallet address
    expect(out).toBe(`Designated wallet address: ${designatedWallet.address}`);
  });

  describe('endorsements', () => {
    it('endorses client', async () => {
      const requester = alice.address;
      const clientAddress = bob.address;

      const out = execCommand(
        'endorse-client',
        ['--mnemonic', mnemonic],
        ['--derivationPath', aliceDerivationPath],
        ['--providerUrl', PROVIDER_URL],
        ['--airnodeRrp', airnodeRrp.address],
        ['--clientAddress', clientAddress]
      );
      expect(out).toBe(`Client address: ${clientAddress}`);

      const endorsed = await admin.requesterToClientAddressToEndorsementStatus(airnodeRrp, requester, clientAddress);
      expect(endorsed).toBe(true);
    });

    it('unendorses client', async () => {
      const requester = alice.address;
      const clientAddress = bob.address;
      airnodeRrp = airnodeRrp.connect(alice);
      await admin.endorseClient(airnodeRrp, clientAddress);

      const isEndorsed = () => admin.requesterToClientAddressToEndorsementStatus(airnodeRrp, requester, clientAddress);

      expect(await isEndorsed()).toBe(true);
      const out = execCommand(
        'unendorse-client',
        ['--mnemonic', mnemonic],
        ['--derivationPath', aliceDerivationPath],
        ['--providerUrl', PROVIDER_URL],
        ['--airnodeRrp', airnodeRrp.address],
        ['--clientAddress', clientAddress]
      );
      expect(out).toBe(`Client address: ${clientAddress}`);
      expect(await isEndorsed()).toBe(false);
    });

    it('check endoresement status', async () => {
      const requester = alice.address;
      const clientAddress = bob.address;

      const checkEndorsement = () =>
        execCommand(
          'get-endorsement-status',
          ['--providerUrl', PROVIDER_URL],
          ['--airnodeRrp', airnodeRrp.address],
          ['--requester', requester],
          ['--clientAddress', clientAddress]
        );

      expect(checkEndorsement()).toBe('Endorsment status: false');
      airnodeRrp = airnodeRrp.connect(alice);
      await admin.endorseClient(airnodeRrp, clientAddress);
      expect(checkEndorsement()).toBe('Endorsment status: true');
    });
  });

  describe('templates', () => {
    const createTemplate = (fileName: string) =>
      execCommand(
        'create-template',
        ['--mnemonic', mnemonic],
        ['--derivationPath', aliceDerivationPath],
        ['--providerUrl', PROVIDER_URL],
        ['--airnodeRrp', airnodeRrp.address],
        ['--templateFilePath', `${__dirname}/../fixtures/${fileName}`]
      );

    it('can create template', async () => {
      const out = createTemplate('template.json');
      expect(out).toMatch(new RegExp('Template ID: 0x\\w+'));
    });

    it('errors out on invalid template file', () => {
      expect(() => createTemplate(`non-existent-file.json`)).toThrow('ENOENT: no such file or directory');
    });

    it('can get template', () => {
      const templateId = createTemplate('template.json').split('Template ID: ')[1];

      const out = execCommand(
        'get-template',
        ['--providerUrl', PROVIDER_URL],
        ['--airnodeRrp', airnodeRrp.address],
        ['--templateId', templateId]
      );

      expect(JSON.parse(out)).toEqual({
        airnodeId: '0x15e7097beac1fd23c0d1e3f5a882a6f99ecbcf2e0c1011d1bd43707c6c0ec717',
        endpointId: '0x2605589dfc93c8f9c35eecdfe1e666c2193df30a8b13e1e0dd72941f59f9064c',
        parameters:
          '0x31537500000000000000000000000000000000000000000000000000000000006e616d653100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a06e616d653200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003e8000000000000000000000000000000000000000000000000000000000000000676616c7565310000000000000000000000000000000000000000000000000000',
      });
    });
  });

  describe('withdrawal', () => {
    let requester: string;
    let airnodeMnemonic: string;
    let airnodeId: string;
    let designatedWallet: ethers.Wallet;
    let destinationWallet: ethers.Wallet;
    const destinationBalance = async () => (await destinationWallet.getBalance()).toString();

    beforeEach(async () => {
      // Prepare for derivation of designated wallet - see test for designated wallet derivation for details
      requester = alice.address;
      const airnodeWallet = ethers.Wallet.createRandom().connect(provider);
      airnodeMnemonic = airnodeWallet.mnemonic.phrase;
      const masterWallet = ethers.Wallet.fromMnemonic(airnodeMnemonic, 'm').connect(provider);
      // Fund the master wallet - which will be used to set the airnode parameters
      await deployer.sendTransaction({
        to: masterWallet.address,
        value: ethers.utils.parseEther('1'),
      });
      airnodeRrp = airnodeRrp.connect(airnodeWallet);
      airnodeId = await admin.setAirnodeParameters(airnodeRrp, []);

      // Derive and fund the designated wallet
      designatedWallet = await deriveDesignatedWallet(airnodeWallet, requester);
      await deployer.sendTransaction({
        to: designatedWallet.address,
        value: ethers.utils.parseEther('1'),
      });

      // Create destination address
      destinationWallet = ethers.Wallet.createRandom().connect(provider);
    });

    it('returns the desintation balance', async () => {
      expect(await destinationBalance()).toBe('0');
    });

    it('can create and fulfill withdrawal request', async () => {
      const requestWithdrawalOutput = execCommand(
        'request-withdrawal',
        ['--mnemonic', mnemonic],
        ['--derivationPath', aliceDerivationPath],
        ['--providerUrl', PROVIDER_URL],
        ['--airnodeRrp', airnodeRrp.address],
        ['--airnodeId', airnodeId],
        ['--requester', requester],
        ['--destination', destinationWallet.address]
      );

      expect(requestWithdrawalOutput).toMatch(new RegExp(`Withdrawal request ID: 0x\\w+`));
      const withdrawalRequestId = requestWithdrawalOutput.split('Withdrawal request ID: ')[1];

      const checkWithdrawalStatus = () =>
        execCommand(
          'check-withdrawal-request',
          ['--providerUrl', PROVIDER_URL],
          ['--airnodeRrp', airnodeRrp.address],
          ['--withdrawalRequestId', withdrawalRequestId]
        );

      expect(checkWithdrawalStatus()).toBe('Withdrawal request is not fulfilled yet');

      airnodeRrp = airnodeRrp.connect(designatedWallet);
      await admin.fulfillWithdrawal(
        airnodeRrp,
        withdrawalRequestId,
        airnodeId,
        requester,
        destinationWallet.address,
        '0.8'
      );
      expect(checkWithdrawalStatus()).toBe('Withdrawn amount: 800000000000000000');
      expect(await destinationBalance()).toBe('800000000000000000');
    });
  });

  it('can set/get airnode parameters', async () => {
    const airnodeWallet = ethers.Wallet.createRandom().connect(provider);
    const airnodeMnemonic = airnodeWallet.mnemonic.phrase;
    const hdNode = ethers.utils.HDNode.fromMnemonic(airnodeMnemonic);
    const xpub = hdNode.neuter().extendedKey;
    const masterWallet = new ethers.Wallet(hdNode.privateKey, provider);
    // Fund the master wallet - which will be used to set the airnode parameters
    await deployer.sendTransaction({
      to: masterWallet.address,
      value: ethers.utils.parseEther('1'),
    });

    const setAirnodeParametersOut = execCommand(
      'set-airnode-parameters',
      ['--mnemonic', airnodeMnemonic],
      ['--derivationPath', 'm'],
      ['--providerUrl', PROVIDER_URL],
      ['--airnodeRrp', airnodeRrp.address],
      ['--authorizersFilePath', `${__dirname}/../fixtures/authorizers.json`]
    );
    expect(setAirnodeParametersOut).toMatch(new RegExp('Airnode ID: \\w+'));

    const airnodeId = setAirnodeParametersOut.split('Airnode ID: ')[1];
    const getAirnodeParametersOut = execCommand(
      'get-airnode-parameters',
      ['--providerUrl', PROVIDER_URL],
      ['--airnodeRrp', airnodeRrp.address],
      ['--airnodeId', airnodeId]
    );

    const json = JSON.parse(getAirnodeParametersOut);
    expect(json).toEqual({
      admin: masterWallet.address,
      authorizers: ['0x0000000000000000000000000000000000000000'],
      xpub,
      blockNumber: expect.anything(),
    });
  });

  it('derives endpoint ID', () => {
    const oisTitle = 'title';
    const endpointName = 'endpoint';

    const out = execCommand('derive-endpoint-id', ['--oisTitle', oisTitle], ['--endpointName', endpointName]);

    const expected = ethers.utils.keccak256(
      ethers.utils.defaultAbiCoder.encode(['string'], [`${oisTitle}_${endpointName}`])
    );
    expect(out).toBe(`Endpoint ID: ${expected}`);
  });

  describe('incorrect usage', () => {
    it('is missing command parameter', () => {
      expect(() =>
        execCommand(
          'endorse-client',
          ['--mnemonic', mnemonic],
          ['--derivationPath', aliceDerivationPath],
          ['--providerUrl', PROVIDER_URL],
          ['--airnodeRrp', airnodeRrp.address]
          // missing ['--clientAddress', clientAddress]
        )
      ).toThrow('Missing required argument: clientAddress');
    });

    it('unknown command', () => {
      expect(() =>
        execCommand('not-existent-command', ['--mnemonic', mnemonic], ['--providerUrl', PROVIDER_URL])
      ).toThrow('Unknown arguments: mnemonic, providerUrl, not-existent-command');
    });
  });
});
