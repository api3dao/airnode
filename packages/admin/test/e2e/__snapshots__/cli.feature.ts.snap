// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI shows help 1`] = `
"cli.js <command>

Commands:
  cli.js create-requester               Creates a requester and returns its
                                        index
  cli.js requester-index-to-admin       Returns the admin of the requesterIndex
  cli.js set-requester-admin            Sets the requester admin
  cli.js derive-designated-wallet       Derives the address of the designated
                                        wallet for a airnode-requester pair
  cli.js endorse-client                 Endorses a client as a requester admin
  cli.js unendorse-client               Unendorses a client as a requester admin
  cli.js get-endorsement-status         Returns the endorsment status for the
                                        given requesterIndex and client
  cli.js create-template                Creates a template and returns its id
  cli.js request-withdrawal             Requests withdrawal from the designated
                                        wallet of an Airnode as a requester
                                        admin
  cli.js check-withdrawal-request       Checks the state of the withdrawal
                                        request
  cli.js fulfill-withdrawal             Fulfils the withdrawal status
  cli.js count-withdrawal-requests      Returns the number of withdrawal
                                        requests for the given requesterIndex
  cli.js set-airnode-parameters         Sets the parameters of an Airnode and
                                        returns its id
  cli.js get-airnode-parameters         Returns the airnode parameters and block
                                        number for the given airnodeId
  cli.js derive-endpoint-id             Derives an endpoint id using the OIS
                                        title and endpoint name
  cli.js client-address-to-no-requests  Returns the number of requests made by
                                        the client
  cli.js get-template                   Returns the template for the given
                                        templateId
  cli.js get-templates                  Returns the templates for the given
                                        templateIds

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]"
`;
